File: src/gmeow_tools/gts/nquads.py
# SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: Apache-2.0
"""The ``gts → nquads`` transform (§14).

Serialises the folded base quads, plus reifier/annotation triples in the RDF 1.2
reifying style (``<reifier> rdf:reifies <<( s p o )>>`` and ``<reifier> p v``).
Inline blobs are externalised by the caller; this module emits the graph text only.
"""

from __future__ import annotations

from gmeow_tools.gts.model import Graph, Term, TermKind

_RDF_REIFIES = "http://www.w3.org/1999/02/22-rdf-syntax-ns#reifies"


def _escape(lex: str) -> str:
    """Escape a literal lexical form for N-Triples."""
    return (
        lex.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    )


def _render(g: Graph, tid: int) -> str:
    """Render a term-id as an N-Triples token."""
    t: Term = g.terms[tid]
    if t.kind is TermKind.IRI:
        return f"<{t.value or ''}>"
    if t.kind is TermKind.BNODE:
        return f"_:{t.value or f'b{tid}'}"
    if t.kind is TermKind.LITERAL:
        lit = f'"{_escape(t.value or "")}"'
        if t.lang is not None:
            return f"{lit}@{t.lang}"
        if t.datatype is not None:
            return f"{lit}^^{_render(g, t.datatype)}"
        return lit  # plain literal == xsd:string (§7.1)
    # quoted triple (RDF 1.2 triple term), resolved through its reifier
    if t.reifier is not None and t.reifier in g.reifiers:
        s, p, o = g.reifiers[t.reifier]
        return f"<<( {_render(g, s)} {_render(g, p)} {_render(g, o)} )>>"
    return "<<( )>>"


def to_nquads(g: Graph) -> str:
    """Serialise a folded :class:`Graph` to N-Quads text."""
    lines: list[str] = []
    for s, p, o, gname in g.quads:
        triple = f"{_render(g, s)} {_render(g, p)} {_render(g, o)}"
        if gname is not None:
            lines.append(f"{triple} {_render(g, gname)} .")
        else:
            lines.append(f"{triple} .")
    for rid, spo in g.reifiers.items():
        quoted = (
            f"<<( {_render(g, spo[0])} {_render(g, spo[1])} {_render(g, spo[2])} )>>"
        )
        lines.append(f"{_render(g, rid)} <{_RDF_REIFIES}> {quoted} .")
    for r, p, v in g.annotations:
        lines.append(f"{_render(g, r)} {_render(g, p)} {_render(g, v)} .")
    return "\n".join(lines) + ("\n" if lines else "")

File: ts/gts/src/nquads.ts
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: Apache-2.0

import { Graph, TermKind } from "./model.js";

const RDF_REIFIES = "http://www.w3.org/1999/02/22-rdf-syntax-ns#reifies";

function escape(lex: string): string {
    let out = "";
    for (const ch of lex) {
        switch (ch) {
            case "\\":
                out += "\\\\";
                break;
            case '"':
                out += '\\"';
                break;
            case "\n":
                out += "\\n";
                break;
            case "\r":
                out += "\\r";
                break;
            case "\t":
                out += "\\t";
                break;
            default: {
                const code = ch.codePointAt(0) ?? 0;
                if (code < 0x20) {
                    out += `\\u${code.toString(16).padStart(4, "0").toUpperCase()}`;
                } else {
                    out += ch;
                }
            }
        }
    }
    return out;
}

function render(g: Graph, tid: number): string {
    if (tid < 0 || tid >= g.terms.length) {
        return `_:out_of_range_${tid}`;
    }
    const t = g.terms[tid];
    switch (t.kind) {
        case TermKind.Iri:
            return `<${t.value}>`;
        case TermKind.Bnode:
            if (t.value !== "") return `_:${t.value}`;
            return `_:b${tid}`;
        case TermKind.Literal: {
            let lit = `"${escape(t.value)}"`;
            if (t.lang) lit += `@${t.lang}`;
            else if (t.datatype !== undefined)
                lit += `^^${render(g, t.datatype)}`;
            return lit;
        }
        case TermKind.Triple:
            if (t.reifier !== undefined) {
                const spo = g.reifier(t.reifier);
                if (spo) {
                    return `<<( ${render(g, spo.s)} ${render(g, spo.p)} ${render(g, spo.o)} )>>`;
                }
            }
            return `_:unbound_triple_${tid}`;
    }
}

/** Serialise a folded Graph to N-Quads text. */
export function toNQuads(g: Graph): string {
    const lines: string[] = [];
    for (const q of g.quads) {
        const triple = `${render(g, q.s)} ${render(g, q.p)} ${render(g, q.o)}`;
        if (q.g !== undefined) {
            lines.push(`${triple} ${render(g, q.g)} .`);
        } else {
            lines.push(`${triple} .`);
        }
    }
    for (const r of g.reifiers) {
        const quoted = `<<( ${render(g, r.spo.s)} ${render(g, r.spo.p)} ${render(g, r.spo.o)} )>>`;
        lines.push(`${render(g, r.rid)} <${RDF_REIFIES}> ${quoted} .`);
    }
    for (const a of g.annotations) {
        lines.push(`${render(g, a.s)} ${render(g, a.p)} ${render(g, a.o)} .`);
    }
    if (lines.length === 0) return "";
    return lines.join("\n") + "\n";
}
